beacon_node/operation_pool/src/lib.rs

```diff
@@ -215,28 +235,121 @@ impl<T: EthSpec> OperationPool<T> {
         self.attestations.read().stats()
     }
 
-    /// Return all valid attestations for the given epoch, for use in max cover.
+    /// Return a vector of aggregate/unaggregate attestations which are maximal cliques
+    /// with resepct to the graph with attestations as vertices and an edge encoding
+    /// compatibility for aggregation.
     #[allow(clippy::too_many_arguments)]
-    fn get_valid_attestations_for_epoch<'a>(
+    #[allow(clippy::type_complexity)]
+    fn get_clique_aggregate_attestations_for_epoch<'a>(
         &'a self,
         checkpoint_key: &'a CheckpointKey,
         all_attestations: &'a AttestationMap<T>,
         state: &'a BeaconState<T>,
-        reward_cache: &'a RewardCache,
-        total_active_balance: u64,
-        validity_filter: impl FnMut(&AttestationRef<'a, T>) -> bool + Send,
+        validity_filter: impl Fn(&CheckpointKey, &CompactAttestationData) -> bool + Send + Sync,
+        num_valid: &AtomicUsize,
+        max_vertices: usize,
         spec: &'a ChainSpec,
-    ) -> impl Iterator<Item = AttMaxCover<'a, T>> + Send {
-        all_attestations
-            .get_attestations(checkpoint_key)
-            .filter(|att| {
-                att.data.slot + spec.min_attestation_inclusion_delay <= state.slot()
-                    && state.slot() <= att.data.slot + T::slots_per_epoch()
+    ) -> Result<Vec<(&CompactAttestationData, Vec<CompactIndexedAttestation<T>>)>, OpPoolError>
+    {
+        let Some(AttestationDataMap {
+            aggregate_attestations,
+            unaggregate_attestations,
+        }) = all_attestations.get_attestation_map(checkpoint_key)
+        else {
+            return Ok(vec![]);
+        };
+        let mut cliques_from_aggregates: Vec<_> = aggregate_attestations
+            .into_par_iter()
+            .filter(|(data, _)| {
+                data.slot + spec.min_attestation_inclusion_delay <= state.slot()
+                    && state.slot() <= data.slot + T::slots_per_epoch()
+            })
+            .map(|(data, aggregates)| {
+                let aggregates: Vec<&CompactIndexedAttestation<T>> = aggregates
+                    .iter()
+                    .filter(|_| validity_filter(checkpoint_key, data))
```

Comment:
I think we're at risk of the deadlock described [in this HackMD](https://hackmd.io/ijQlqOdqSGaWmIo6zMVV-A?view#The-deadlock-in-Lighthouse) (see also https://github.com/sigp/lighthouse/pull/1562).


The `validity_filter` function can try to access the fork choice read lock:

https://github.com/sigp/lighthouse/blob/051c3e842fb5553d3ef3370560b74072c6e07cc6/beacon_node/beacon_chain/src/beacon_chain.rs#L2297

That means this function would behave the same as the `rest_api::validator::publish_attestations` function I describe in the HackMD.

In terms of a workaround, I'm not fully across this PR but I feel like the parallel functionality is most important for the second `map` fn (the one which calls `bron_kerbosch`). So perhaps the `filter` and first `map` could happen in a good ol' fashioned serial iter?

Reply:
> Without holding an exclusive lock/mutex while spawning I think we are safe.

I think that we do that here:

https://github.com/sigp/lighthouse/blob/2bc9115a94bfcf1269f83f0a9e7a1321b2a7f70c/beacon_node/beacon_chain/src/beacon_chain.rs#L1500-L1502

We take a write-lock on fork choice, then we call `on_block` which might read a state from the DB which might involve rayon-parallelised tree hashing.

So the deadlock would be:

1. We take the fork choice write lock in `on_block` but don't yet trigger tree hashing.
1. We run the `par_iter` in this PR, which fills all the rayon threads with functions waiting on the fork choice read lock.
1. `on_block` (via the `store`) tries to start tree hashing but it can't start because all the rayon threads are busy.
1. We are deadlocked because we need `on_block` to finish so we can drop the fork choice write-lock.

Reply:
Oh bugger, I think that could deadlock even with our current code then.

1. Never hold a lock while spawning Rayon tasks, or
2. Never obtain a lock in a Rayon task that could be held over a spawn

I have a feeling that (1) is better in some way, but need to sleep on it.

Reply:
Actually does `on_block` load states? I can't see it at a quick glance. I also don't think loading a state does any tree hashing (although it will in tree-states land, so definitely something to keep an eye on).

Reply:
Ah yeah, here:

https://github.com/sigp/lighthouse/blob/441fc1691b69f9edc4bbdc6665f3efab16265c9b/beacon_node/beacon_chain/src/beacon_fork_choice_store.rs#L328-L334

Reply:
If I get some time today or tomorrow I'll try writing a test with `hiatus` that deadlocks on `unstable`.

Reply:
After _a lot_ of mucking around I got block production & tree hashing to deadlock on `unstable`, sort of. The branch is here: https://github.com/michaelsproul/lighthouse/commit/5e253dee7bda9b22ae537af666122707aebd8cf7

It was really hard to get all the bits of line up, so there are 7 separate `hiatus` breakpoints. I also had to cheat and just force tree hashing in `import_block` because I ran out of patience to get the tree hashing and state load in `on_block` to trigger. The main problem is that the attestations were triggering the justified checkpoint update before I could trigger it with `on_block`. I think using the mainnet spec would make this a bit easier as we wouldn't have to justify on the epoch boundary (as we do with minimal).

I think this deadlock (on `unstable`) hasn't been observed in practice for several reasons:

1. It requires block production and block import to be running concurrently. This doesn't happen particularly often, but isn't _that_ unlikely.
2. It requires the balances cache in fork choice to miss. This almost never happens in recent versions of Lighthouse.
3. It requires a machine with ~2 cores, because we only spawn two op pool tasks using `rayon::join`. If there are more than 2 Rayon threads in the pool then the tree hashing can make progress and drop the fork choice lock, which unblocks the op pool tasks.

In my example you can see the deadlock happen with:

```
RAYON_NUM_THREADS=2 FORK_NAME=capella cargo test --release --features fork_from_env -- deadlock_block_production --nocapture
```

If you bump the threadpool size to 3 then the deadlock doesn't happen (due to condition 3).

```
RAYON_NUM_THREADS=3 FORK_NAME=capella cargo test --release --features fork_from_env -- deadlock_block_production --nocapture
```

I haven't tried deadlocking Geemo's branch yet, but intuitively we make the deadlock more likely by spawning more threads for the attestation data (condition 3 no longer applies). Still, we are protected by the unlikeliness of condition 1 & condition 2 occurring simultaneously.

I need to think about it more.